### PR TITLE
Song relations

### PR DESCRIPTION
**Pull request by ridhoq**
you guys silly? i'm still gonna send it

https://www.youtube.com/watch?v=RSuLFvalhnQ

### CODE REVIEW

File: app/api/song_relations.py [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+from ..models import SongRelation, SongRelationVote
+from .. import db, auth
+from . import api
+from .errors import bad_request, route_not_found
+
+@api.route('/song_relations/', methods=['POST'])
+@auth.login_required
+def new_song_relation():
+    # check if json
+    if request.headers['content_type'] == 'application/json':
```
Comment (austinpray) [inline app/api/song_relations.py]: Damn dude stick this in a middleware haha

File: app/api/song_relations.py [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+                return bad_request(message)
+
+            song_relation = SongRelation(song1_id, song2_id, g.current_user)
+            db.session.add(song_relation)
+            db.session.commit()
+            vote_song_relation_helper(song_relation.id)
+            return make_response(jsonify(song_relation.to_json(g.current_user)), 200)
+
+        except Exception as ex:
+            template = "An exception of type {0} occured. Arguments:\n{1!r}"
```
Comment (austinpray) [inline app/api/song_relations.py]: Can use the % operator here

File: app/api/song_relations.py [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+    return make_response(jsonify(song_relation.to_json(g.current_user)), 200)
+
+@api.route('/song_relations/<int:id>/vote', methods=['DELETE'])
+@auth.login_required
+def delete_vote_song_relation(id):
+    song_relation_vote = SongRelationVote.query.filter_by(song_relation_id=id, user_id=g.current_user.id).first()
+    if song_relation_vote is None:
+        message = 'this user has not voted for this relation'
+        return bad_request(message)
+    db.session.delete(song_relation_vote)
```
Comment (austinpray) [inline app/api/song_relations.py]: Makes me nervous. What if song_relation_vote accidentally selects a bunch of votes cause the filter params are null? Why not just specify the id explicitly?

File: app/models.py [hunk 4]
```diff
@@ -41,6 +41,32 @@ def to_json(self):
...
+            join song_relation_votes as srv
+                on srv.song_relation_id = sr.id
+            join songs as rel
+                on (rel.id = sr.song1_id and rel.id <> sr.song2_id)
+                or (rel.id = sr.song2_id and rel.id <> sr.song1_id)
+            where s.id = {0} and rel.id <> {0}
+            group by rel.id, rel.title, rel.url, rel.artist, rel.created, rel.user_id
+            order by vote_count desc
+            limit {1}
+        '''.format(self.id, top)
```
Comment (austinpray) [inline app/models.py]: Rofl dude this is a sql injection vulnerability. Top comes from query args and you don't ever check if it's an int.

File: app/api/songs.py [hunk 5]
```diff
@@ -60,5 +60,14 @@ def new_song():
         message = 'that aint json'
         return bad_request(message)
 
+@api.route('/songs/<int:id>/related')
+def get_song_relations(id):
+    top = request.args.get('top')
```
Comment (austinpray) [inline app/api/songs.py]: Sql injection vuln. You never check that this is an int.

File: app/models.py [hunk 6]
```diff
@@ -97,3 +124,75 @@ def to_json(self):
...
+
+    def to_json(self, user):
+        json_song_relation = {
+            'id': self.id,
+            'song1': Song.query.filter_by(id=self.song1_id).first().to_json(),
+            'song2': Song.query.filter_by(id=self.song2_id).first().to_json(),
+            'has_voted': self.has_voted(user),
+            'vote_count': SongRelationVote.query.filter_by(song_relation_id=self.id).count(),
+            'created': self.created,
+            'created_by': User.query.filter_by(id=self.user_id).first().to_json()
```
Comment (austinpray) [inline app/models.py]: Reeee usually JSON is camelCase

File: app/models.py (hunk 4 above)
Comment (ridhoq) [inline app/models.py]: lmao

**Pull request merged**